File: backend/app/services/prospect_activity.py
"""Estado operativo visible de un prospecto en listas de campaña."""

from __future__ import annotations

from typing import Any

from app.models.prospect import Prospect


def compute_prospect_activity(prospect: Prospect) -> dict[str, Any]:
    """
    Qué está pasando ahora con este contacto (para UI al costado del nombre).

    Prioridad: enrich → pausa → LinkedIn conexión → LinkedIn mensaje → WhatsApp →
    secuencia / email → idle.
    """
    enrich = (getattr(prospect, "channel_enrich_status", None) or "").strip().lower()
    if enrich == "searching":
        msg = (getattr(prospect, "channel_enrich_message", None) or "").strip()
        return {
            "code": "enriching",
            "label": msg or "Buscando información de canales…",
            "deadline_at": getattr(prospect, "channel_enrich_deadline_at", None),
            "tone": "search",
        }

    if bool(getattr(prospect, "sequence_paused", False)):
        return {
            "code": "paused",
            "label": "Secuencia pausada",
            "deadline_at": None,
            "tone": "muted",
        }

    status = (getattr(prospect, "status", None) or "").strip().lower()
    if status in ("meeting_booked", "won"):
        return {
            "code": "meeting",
            "label": "Reunión agendada",
            "deadline_at": None,
            "tone": "ok",
        }
    if status == "not_interested":
        return {
            "code": "closed",
            "label": "No interesado",
            "deadline_at": None,
            "tone": "muted",
        }

    started = getattr(prospect, "sequence_started_at", None) is not None
    if not started:
        if enrich == "timed_out":
            return {
                "code": "enrich_timeout",
                "label": "Sin datos extra · listo para iniciar",
                "deadline_at": None,
                "tone": "muted",
            }
        return {
            "code": "idle",
            "label": "Guardado · esperando inicio",
            "deadline_at": None,
            "tone": "muted",
        }

    conn = (getattr(prospect, "linkedin_connection_status", None) or "none").strip().lower()
    if conn in ("invite_pending", "invite_sent", "checking", "check_queued", "check_failed"):
        if conn == "check_failed":
            label = "No se pudo verificar LinkedIn"
        elif conn in ("checking", "check_queued"):
            label = "Verificando si es contacto en LinkedIn"
        else:
            label = "Esperando conexión en LinkedIn"
        return {
            "code": "linkedin_connect",
            "label": label,
            "deadline_at": None,
            "tone": "wait",
        }

    li_assist = (getattr(prospect, "linkedin_assist_status", None) or "").strip().lower()
    li_sent = getattr(prospect, "linkedin_sdr_marked_sent_at", None) is not None
    if conn == "connected" and not li_sent and li_assist in (
        "suggested",
        "prepared",
        "opened",
        "queued",
    ):
        return {
            "code": "linkedin_message",
            "label": "Esperando envío de mensaje LinkedIn",
            "deadline_at": None,
            "tone": "wait",
        }

    wa_assist = (getattr(prospect, "whatsapp_assist_status", None) or "").strip().lower()
    if wa_assist in ("suggested", "prepared", "opened", "queued"):
        return {
            "code": "whatsapp_message",
            "label": "Esperando envío de WhatsApp",
            "deadline_at": None,
            "tone": "wait",
        }

    # Labels de secuencia si están
    current = (getattr(prospect, "sequence_current_label", None) or "").strip()
    next_lbl = (getattr(prospect, "next_touch_label", None) or "").strip()
    day_lbl = (getattr(prospect, "sequence_current_day_label", None) or "").strip()
    if (current and "gmail" in current.lower()) or (
        next_lbl and "email" in next_lbl.lower()
    ):
        return {
            "code": "email",
            "label": current or next_lbl or "Enviando / pendiente Gmail",
            "deadline_at": None,
            "tone": "active",
        }
    if current or next_lbl or day_lbl:
        return {
            "code": "sequence",
            "label": current or next_lbl or day_lbl,
            "deadline_at": None,
            "tone": "active",
        }

    group = (getattr(prospect, "sequence_group", None) or "").strip().lower()
    if group == "encajonado":
        return {
            "code": "boxed",
            "label": "Encajonado · esperando respuesta",
            "deadline_at": None,
            "tone": "wait",
        }
    if group == "postergado":
        return {
            "code": "deferred",
            "label": "Postergado",
            "deadline_at": None,
            "tone": "muted",
        }

    # Nunca devolver «Iniciando secuencia…».
    return {
        "code": "none",
        "label": "",
        "deadline_at": None,
        "tone": "muted",
    }

File: frontend/src/utils/prospectActivity.js
/**
 * Estado operativo de un prospecto (lista de campaña / secuencia).
 * @param {object} p
 * @returns {{ code: string, label: string, tone: string, deadlineAt?: string|null, showCountdown?: boolean }}
 */
export function resolveProspectActivity(p) {
  if (!p || typeof p !== 'object') {
    return { code: 'unknown', label: '—', tone: 'muted' }
  }

  // Preferir campos del backend si vienen (label vacío = sin badge de estado).
  if (p.activity_label != null && String(p.activity_code || '') !== '') {
    const label = String(p.activity_label || '').trim()
    if (!label || p.activity_code === 'none') {
      // seguir con heurística local / null
    } else {
      return {
        code: p.activity_code || 'custom',
        label,
        tone: p.activity_tone || 'muted',
        deadlineAt: p.channel_enrich_deadline_at || p.activity_deadline_at || null,
        showCountdown:
          String(p.activity_code || '') === 'enriching' ||
          String(p.channel_enrich_status || '').toLowerCase() === 'searching',
      }
    }
  }

  const enrich = String(p.channel_enrich_status || '').toLowerCase()
  if (enrich === 'searching') {
    return {
      code: 'enriching',
      label: p.channel_enrich_message || 'Buscando información de canales…',
      tone: 'search',
      deadlineAt: p.channel_enrich_deadline_at || null,
      showCountdown: true,
    }
  }

  if (p.sequence_paused) {
    return { code: 'paused', label: 'Secuencia pausada', tone: 'muted' }
  }

  const status = String(p.status || '').toLowerCase()
  if (status === 'meeting_booked' || status === 'won') {
    return { code: 'meeting', label: 'Reunión agendada', tone: 'ok' }
  }
  if (status === 'not_interested') {
    return { code: 'closed', label: 'No interesado', tone: 'muted' }
  }

  if (!p.sequence_started_at) {
    return { code: 'idle', label: 'Guardado · esperando inicio', tone: 'muted' }
  }

  const conn = String(p.linkedin_connection_status || 'none').toLowerCase()
  if (['invite_pending', 'invite_sent', 'checking', 'check_queued', 'check_failed'].includes(conn)) {
    return {
      code: 'linkedin_connect',
      label:
        conn === 'check_failed'
          ? 'No se pudo verificar LinkedIn'
          : conn === 'checking' || conn === 'check_queued'
            ? 'Verificando si es contacto en LinkedIn'
            : 'Esperando conexión en LinkedIn',
      tone: 'wait',
    }
  }

  const liAssist = String(p.linkedin_assist_status || '').toLowerCase()
  const liSent = Boolean(p.linkedin_sdr_marked_sent_at)
  if (
    conn === 'connected' &&
    !liSent &&
    ['suggested', 'prepared', 'opened', 'queued'].includes(liAssist)
  ) {
    return { code: 'linkedin_message', label: 'Esperando envío de mensaje LinkedIn', tone: 'wait' }
  }

  const waAssist = String(p.whatsapp_assist_status || '').toLowerCase()
  if (['suggested', 'prepared', 'opened', 'queued'].includes(waAssist)) {
    return { code: 'whatsapp_message', label: 'Esperando envío de WhatsApp', tone: 'wait' }
  }

  const current = String(p.sequence_current_label || '').trim()
  const nextLbl = String(p.next_touch_label || '').trim()
  if (/gmail|email|correo/i.test(`${current} ${nextLbl}`)) {
    return {
      code: 'email',
      label: current || nextLbl || 'Enviando / pendiente Gmail',
      tone: 'active',
    }
  }
  if (current || nextLbl) {
    return { code: 'sequence', label: current || nextLbl, tone: 'active' }
  }

  const group = String(p.sequence_group || '').toLowerCase()
  if (group === 'encajonado') {
    return { code: 'boxed', label: 'Encajonado · esperando respuesta', tone: 'wait' }
  }
  if (group === 'postergado') {
    return { code: 'deferred', label: 'Postergado', tone: 'muted' }
  }

  // Nunca mostrar «Iniciando secuencia…».
  return null
}
